### ISSUE TITLE
Allow printing stacktrace on uncaught main thread exception

### ISSUE THREAD

**Issue by lfielke** [getsentry/sentry-java#1077]
I'd like to be able to use Sentry and still have the Java default behavior of printing a stacktrace when the main thread throws an uncaught exception.

I'm using Sentry on a server project with a `main` method that does some initialization and reading configuration files before starting the service proper. The code is written in a fail-fast way, and simply propagates exceptions letting the program crash. Before adding Sentry, for example if the config file has a string where a number is expected `NumberFormatException` will be thrown and cause the program to exit and a stacktrace will be printed showing the exact place the error occurred. After adding Sentry, the stacktrace is no longer printed (because `UncaughtExceptionHandlerIntegration` calls `setDefaultUncaughtExceptionHandler`, overriding the built in standard library call to `printStackTrace`). This means that failures are really jarring when running the program interactively, because it just terminates with no error message and you're left wondering what happened.

I would like a way to be able to re-enable printing the stacktrace to stderr for the main thread. Alternately (matching the standard library behavior) a way that prints it for any thread in the `ThreadGroup` with no parent (the "system" `ThreadGroup`).

I have discovered a workaround that works enough for my purposes, but it's not pretty because there's no easy way to access the `UncaughtExceptionHandlerIntegration` instance:

```java
Sentry.init(options -> {
    options.setDsn("https://abc123xxx.ingest.sentry.io/123456");
    // Other options as per usual...

    // Find the instance of the UncaughtExceptionHandlerIntegration
    UncaughtExceptionHandlerIntegration uehi = (UncaughtExceptionHandlerIntegration) options.getIntegrations()
            .stream()
            .filter(integration -> integration instanceof UncaughtExceptionHandlerIntegration)
            .findFirst()
            .orElseThrow(() -> new IllegalStateException("Expected to find a Sentry UncaughtExceptionHandlerIntegration"));
    // Install our own UncaughtExceptionHandler just for the calling thread (main thread) that prints the stacktrace then chains to Sentry.            
    Thread.currentThread().setUncaughtExceptionHandler((t, e) -> {
                System.err.print("Exception in thread \"" + t.getName() + "\" ");
                e.printStackTrace(System.err);
                uehi.uncaughtException(t, e);
            }
    );
});
```

I also tried a working around this using a different approach, which was to disable the exception integration `setEnableUncaughtExceptionHandler(false)`, however this seems to be a global setting; there's no way just to disable it for the main thread. I didn't want to write my own DefaultUncaughtExceptionHandler that forwarded to Sentry, because that seemed like for work than my (hacky!) workaround above.

As a point of comparison, BugSnag (another error tracking service) explicitly emulates the stacktrace printing behavior in their ExceptionHandler, without giving users a choice: https://github.com/bugsnag/bugsnag-java/blob/1ff0be9af090b0e4913160249c0952a1ff934d31/bugsnag/src/main/java/com/bugsnag/ExceptionHandler.java#L66-L68

**Comment by marandaneto** [comment]
@lfielke hey thanks for raising this, for server Apps, that makes sense.

what do you think about something like this?

```
      if (defaultExceptionHandler != null) {
        options.getLogger().log(SentryLevel.INFO, "Invoking inner uncaught exception handler.");
        defaultExceptionHandler.uncaughtException(thread, thrown);
      } else {
        if (options.printUncaughtStackTrace) {
          options.getLogger().log(SentryLevel.INFO, "Exception in thread: %s, StackTrace: ", thread.getName(), thrown.printStackTrace());
        }
      }
```

just pseudocode though.

This would only print if there's not a `defaultExceptionHandler`, because if it does,. then the client should do whatever.

I confess this was only thought for an Android env at the beginning where you don't have access to the console output, that's why I'd add an option for that `printUncaughtStackTrace` or similar, it won't make sense for Mobile apps.

**Comment by francisdb** [comment]
I can confirm errors are swallowed by sentry. If for some reason sentry can't send the error it's lost forever without trace.

In 1.x there used to be an option `uncaught.handler.enabled=false` which is no longer available. We prefer to set up our own handler that logs the error through logback that in turn sends errors to sentry.

**Comment by marandaneto** [comment]
@francisdb not sure If I follow your comment.

Sentry calls the default `DefaultUncaughtExceptionHandler` as shown in the above code snippet (after calling Sentry.capture...).

`uncaught.handler.enabled=false` is pretty much the same as `setEnableUncaughtExceptionHandler(false)`, could you elaborate?

**Comment by francisdb** [comment]
By default sentry changes the default jvm behaviour by making it no longer log stack traces, with a possibility that it can't send the error to sentry.

If we make sure we have our own handler configured before any logging starts up, and that handler writes the error to logback, we end up with duplicate entries in sentry. As the logback appender with log the error and the sentry exception handler will log it.

So on 1.x we disabled the handler using the `sentry.properties` file

Also we don't want compile time dependencies and configure everything using the env/properties/system properties. So calling `setEnableUncaughtExceptionHandler` is not an option.

**Comment by francisdb** [comment]
But maybe I should create a new issue for our specific case?

**Comment by marandaneto** [comment]
@francisdb makes sense, pretty disconnected from this one (at least the setting from external sources)

**Comment by lfielke** [comment]
@marandaneto

That logic seems okay to me, although will this require `options.setLogger` to set a logger to cause anything to be printed? (I think the default is that Sentry doesn't output its own logs anywhere). I think it would be easier to understand if setting `setPrintUncaughtStackTrace(true)` was enough to make it print without also having to setup Sentry debug logging.

Also, I'd perhaps suggest a `FATAL` level instead of `INFO`, as it's an uncaught exception that will cause the thread to terminate.

I understand the different requirements of Android versus server too, so I like the idea of `printUncaughtStackTrace` option.

> ```
>       if (defaultExceptionHandler != null) {
>         options.getLogger().log(SentryLevel.INFO, "Invoking inner uncaught exception handler.");
>         defaultExceptionHandler.uncaughtException(thread, thrown);
>       } else {
>         if (options.printUncaughtStackTrace) {
>           options.getLogger().log(SentryLevel.INFO, "Exception in thread: %s, StackTrace: ", thread.getName(), thrown.printStackTrace());
>         }
>       }
> ```

**Comment by marandaneto** [comment]
yeah makes sense as we don't set a logger if `debug` is not enabled, so I'd go for `thrown.printStackTrace()` directly if `printUncaughtStackTrace` is enabled.

**Comment by bruno-garcia** [comment]
Makes sense to me. Ideally without the options or at least consider that if there's a default handler before we hook, the options won't take effect so it could be misleading

**Comment by Paxa** [comment]
+1 for printUncaughtStackTrace
Also callback function would be nice, to customize how exceptions are printed (but not critical)
